Adapt video-application layout at different sizes

lyzadanger opened the pull request:
This PR adds a mechanism for adapting video-annotation application layout at different app-container sizes and applies some initial layout adaptation. It also adds a new “top bar” with branding and the search/filtering input. This top bar extends the top bar of the client’s sidebar, visually.

While the mechanism for adapting layout is solid here, the applied layout/design (e.g. breakpoints) should be considered first-pass only. There are still some details to iron out. Expect at least one more iteration of the breakpoints/adaptations themselves.

NB: The red loading bar you see here is being [tracked separately](https://github.com/hypothesis/via/issues/991). 

A few screenshots:

<img width="612" alt="Screen Shot 2023-06-15 at 3 18 04 PM" src="https://github.com/hypothesis/via/assets/439947/cd0d8fb8-a4ce-4447-b9d5-b4a5f4ca5cd6">

<img width="612" alt="Screen Shot 2023-06-15 at 3 19 32 PM" src="https://github.com/hypothesis/via/assets/439947/f4b70eb6-c370-4183-9016-b77bde442856">

<img width="1279" alt="Screen Shot 2023-06-15 at 3 21 10 PM" src="https://github.com/hypothesis/via/assets/439947/f9b13bef-b075-452b-8d4b-e7285c89a0d6">

<img width="1599" alt="Screen Shot 2023-06-15 at 3 21 15 PM" src="https://github.com/hypothesis/via/assets/439947/77508fe2-368f-480c-af8c-9d4b6c0d54a1">

### Known awkwardness

There are viewport widths around ~900px where the layout is awkward with the sidebar open. This is a band of widths at which side-by-side is not applied. We may choose to come at this by adjusting the minimum-width threshold of side-by-side for this application. Or we may still want to solve it from the video-app side. In any case, I will track it separately.

<img width="1021" alt="Screen Shot 2023-06-15 at 3 21 02 PM" src="https://github.com/hypothesis/via/assets/439947/8e2463e6-50e1-4b06-89dd-8bf48d2b4afa">

Part of https://github.com/hypothesis/via/issues/953

lyzadanger committed:
Add `wait` test utilities

lyzadanger committed:
Add `useAppLayout` hook

Add a hook that uses `ResizeObserver` to determine relative size of
provided `appContainer`.

lyzadanger committed:
Add `viewbox` attribute to SVGs

This ensures that they won't be cut off if rendered at smaller sizes.

lyzadanger committed:
Adapt layout at different app-container sizes

lyzadanger commented on via/static/scripts/test-util/wait.ts:
I know we've been talking about centralizing these test utils in a shared package, but I needed them nooowwwww :)

lyzadanger commented on via/static/scripts/video_player/components/VideoPlayerApp.tsx:
```diff
@@ -158,46 +169,40 @@ export default function VideoPlayerApp({
-        />
-      </div>
+    <div
+      data-testid="app-container"
```
I acknowledge that this component's layout is getting complicated. To help a tiny bit for now, I've attempted to comment the heck out of styling and add `data-testid`s to elements.

lyzadanger commented on via/static/scripts/video_player/components/VideoPlayerApp.tsx:
```diff
@@ -158,46 +169,40 @@ export default function VideoPlayerApp({
-          )}
-          data-testid="search-bar"
-        >
+        <LogoIcon />
```
Presumably this should link to somewhere and have a `title`, etc. We'll get there.

robertknight replied:
Yes. See the logo link on eg. https://hypothes.is/search.

lyzadanger commented on via/static/scripts/video_player/components/VideoPlayerApp.tsx:
```diff
@@ -158,46 +169,40 @@ export default function VideoPlayerApp({
+              // TODO: Temporary prevention of sidebar controls overlapping
+              'mr-[22px]',
+              {
+                // Adapt width to match width of transcript
```
If I can find a straightforward way to abstract some of this away in a future iteration, I will. I tried a couple of abortive things today but felt in danger of imposing premature abstraction. So for now please have patience with verbose styling until we get into more of a "clean-up" mode.

lyzadanger commented on via/static/scripts/video_player/components/icons/Pause.tsx:
```diff
@@ -14,6 +14,7 @@ export default function PauseIcon(props: PauseIconProps) {
       height="16"
       fill="none"
       data-component="PauseIcon"
+      viewBox="0 0 16 16"
```
This prevents the SVG from getting clipped when rendered at smaller sizes.

robertknight replied:
This is an auto-generated component, using the script from the frontend-shared repo. Can you update the source SVG and regenerate instead. I thought I added a README.md in `components/icons` to explain this, but it looks like I forgot.

lyzadanger replied:
I'm also adding a README

lyzadanger commented on via/static/scripts/video_player/hooks/test/use-app-layout-test.js:
```diff
@@ -0,0 +1,82 @@
+
+  function renderComponent() {
+    act(() => {
+      render(<FakeComponent />, container);
```
NB: Using `preact`'s `render` here. I could not for the life of me find a test structure that worked with enzyme's `mount`.

robertknight replied:
I think I see the issue you ran into. There are two things that need to be done to make this work using our normal way of writing tests:

1. Sync the Enzyme _root_ wrapper with the DOM when polling in `waitFor`, using `wrapper.update()`. Note this doesn't work on a child wrapper.
2. Re-fetch the child wrapper from the root after updating

```diff
diff --git a/via/static/scripts/video_player/hooks/test/use-app-layout-test.js b/via/static/scripts/video_player/hooks/test/use-app-layout-test.js
index 698fa9e..3b79835 100644
--- a/via/static/scripts/video_player/hooks/test/use-app-layout-test.js
+++ b/via/static/scripts/video_player/hooks/test/use-app-layout-test.js
@@ -1,22 +1,17 @@
-import { render } from 'preact';
+import { mount } from 'enzyme';
 import { useRef } from 'preact/hooks';
-import { act } from 'preact/test-utils';
 
 import { waitFor } from '../../../test-util/wait';
 import { useAppLayout } from '../use-app-layout';
 
 describe('useAppLayout', () => {
   let container;
-  let appSize;
-
-  function findElementByTestId(testId) {
-    return container.querySelector(`[data-testid=${testId}]`);
-  }
+  let wrappers;
 
   // Create a fake component to mount in tests that uses the hook
   function FakeComponent() {
     const myRef = useRef();
-    appSize = useAppLayout(myRef);
+    const appSize = useAppLayout(myRef);
     return (
       <div
         ref={myRef}
@@ -30,20 +25,21 @@ describe('useAppLayout', () => {
   }
 
   function renderComponent() {
-    act(() => {
-      render(<FakeComponent />, container);
-    });
-
-    return findElementByTestId('appContainer');
+    const wrapper = mount(<FakeComponent />, { attachTo: container });
+    wrappers.push(wrapper);
+    return wrapper;
   }
 
   beforeEach(() => {
     container = document.createElement('div');
     container.setAttribute('id', 'container');
     document.body.append(container);
+
+    wrappers = [];
   });
 
   afterEach(() => {
+    wrappers.forEach(w => w.unmount());
     container.remove();
   });
 
@@ -55,17 +51,19 @@ describe('useAppLayout', () => {
   ].forEach(({ containerWidth, expected }) => {
     it('should provide relative app size when component is rendered', () => {
       container.style.width = `${containerWidth}px`;
-
-      renderComponent();
-      assert.equal(appSize, expected);
+      const appContainerEl = renderComponent().find(
+        '[data-testid="appContainer"]'
+      );
+      assert.equal(appContainerEl.prop('data-app-size'), expected);
     });
   });
 
   it('should update app size if relative size of container changes', async () => {
     container.style.width = '200px';
 
-    const appContainerEl = renderComponent();
-    assert.equal(appSize, 'sm');
+    const wrapper = renderComponent();
+    const appContainerEl = wrapper.find('[data-testid="appContainer"]');
+    assert.equal(appContainerEl.prop('data-app-size'), 'sm');
 
     const steps = [
       [800, 'md'],
@@ -75,8 +73,13 @@ describe('useAppLayout', () => {
       [400, 'sm'],
     ];
     for (const [width, expected] of steps) {
-      appContainerEl.style.width = `${width}px`;
-      await waitFor(() => appSize === expected, /* timeout */ 50);
+      container.style.width = `${width}px`;
+
+      await waitFor(() => {
+        wrapper.update();
+        const appContainerEl = wrapper.find('[data-testid="appContainer"]');
+        return appContainerEl.prop('data-app-size') === expected;
+      }, /* timeout */ 50);
     }
   });
 });
```

lyzadanger commented on via/static/scripts/video_player/hooks/test/use-app-layout-test.js:
```diff
@@ -0,0 +1,82 @@
+    ];
+    for (const [width, expected] of steps) {
+      appContainerEl.style.width = `${width}px`;
+      await waitFor(() => appSize === expected, /* timeout */ 50);
```
These tests regularly exceed the default `10ms` timeout locally. Is `50` enough leeway? I hope so? Advice welcome.

robertknight replied:
I'm not familiar with `ResizeObserver`'s reporting frequency. It might be synced to `requestAnimationFrame` in which case it would likely be 16ms (60 frames/sec). 50ms might be OK.

lyzadanger replied:
The `requestAnimationFrame` involvement here hadn't occurred to me. Thanks for the note.

robertknight commented on via/static/scripts/test-util/wait.ts:
```diff
@@ -0,0 +1,35 @@
+  condition: () => T,
+  timeout = 10,
+  what = condition.toString()
+) {
```
```suggestion
): Promise<T> {
```

robertknight commented on via/static/scripts/test-util/wait.ts:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Wait for a condition to evaluate to a truthy value.
+ *
+ * @return {Promise<T>} - Result of the `condition` function
```
```suggestion
 * @return Result of the `condition` function
```

robertknight commented on via/static/scripts/video_player/components/VideoPlayerApp.tsx:
```diff
@@ -158,46 +169,40 @@ export default function VideoPlayerApp({
-          'mr-[30px]'
+          // Match height of sidebar's top-bar
+          'h-[40px] min-h-[40px]',
+          // Background is one shade darker than sidebar top-bar
```
Maybe separate the layout and color classes here to make it more obvious what this comment is referring to.

robertknight commented on via/static/scripts/video_player/components/VideoPlayerApp.tsx:
```diff
@@ -209,21 +214,87 @@ export default function VideoPlayerApp({
+          <div
+            data-testid="transcript-controls"
+            className={classnames(
+              // Echo height of top-bar
```
Can you expand this to clarify what "top-bar" is referring to here.

lyzadanger replied:
It should just be worded "Same height as top-bar". I'm updating the comment.

robertknight commented on via/static/scripts/video_player/components/icons/Play.tsx:
```diff
@@ -14,6 +14,7 @@ export default function PlayIcon(props: PlayIconProps) {
       height="16"
       fill="none"
       data-component="PlayIcon"
+      viewBox="0 0 16 16"
```
See comment about icon component generation for Pause icon.

robertknight commented on via/static/scripts/video_player/hooks/use-app-layout.ts:
```diff
@@ -0,0 +1,55 @@
+ * (`body`) width that needs to be queried, not the viewport. We could replace
+ * this with container queries once browser support is sufficient.
+ */
+export function useAppLayout(appContainer: RefObject<HTMLDivElement | null>) {
```
I suggest to explicitly specify the return type here for the reader, and also export the `AppSize` type.

robertknight commented on via/static/scripts/video_player/hooks/use-app-layout.ts:
```diff
@@ -0,0 +1,55 @@
+      onSizeChanged(element.clientWidth, element.clientHeight)
+    );
+    observer.observe(element);
+    return () => observer.disconnect();
```
The `observeElementSize` helper seems redundant to me if we're only using `ResizeObserver` internally. It made more sense when it was abstracting over different approaches.

lyzadanger replied:
Good point. I kind of looked past it since it was copied code. I'll unfurl it.

robertknight commented on via/static/scripts/video_player/hooks/use-app-layout.ts:
```diff
@@ -0,0 +1,55 @@
+    const cleanup = observeElementSize(appContainer.current!, updateWidth);
+    updateWidth();
+    return cleanup;
+  }, [updateWidth, appContainer]);
```
This layout effect will never re-run if the `appContainer.current` value changes, because the hook is watching the ref object rather than the `current` property.

Some options:

- Accept this limitation, but call it out in comments. I think this is the easiest option.
- Re-work hook to take an observable ref of some sort and then do the work to handle changes in the value internally. This effort is not justified at present IMO, unless we can find a really easy way of doing it.

robertknight reviewed (commented):
The general shape of this makes sense. I left some notes on the tests, icons and a couple of other things. I understand the layouts are not finalized yet. The one major comment I have on it is that I think we should make more space for the transcript in stacked layout and also have the transcript search field nearer the actual transcript in that case. It seems weird to have the video in the middle between them.

lyzadanger commented:
Updated:

* Icon components regenerated, and README added
* Updated return types in `wait`
* Explicit return type in hook; export `AppSize` type; document limitation with `appContainer.current`; unrolled internal function
* Hypothesis icon linked
* Updated styling comments in `VideoPlayerApp`
* Applied suggested test changes (and wrapped head around them)

Deferring, but tracked per suggestion:

#993
#994

lyzadanger commented on via/static/scripts/video_player/hooks/use-app-layout.ts:
```diff
@@ -0,0 +1,54 @@
+
+  useLayoutEffect(() => {
+    const element = appContainer.current;
+    /* istanbul ignore if */
+    if (!element) {
+      return () => {};
+    }
```
This state should never occur. Should I remove this guard?

robertknight replied:
I would probably leave as is, or maybe log a warning. You could make an argument that having `appContainer.current` be null is likely a mistake and so the code should fail loudly. On the other hand, the hook is slightly more flexible in its current state.

lyzadanger commented on via/static/scripts/video_player/components/VideoPlayerApp.tsx:
```diff
@@ -158,46 +169,48 @@ export default function VideoPlayerApp({
-          data-testid="search-bar"
+        <a
+          href="https://web.hypothes.is"
+          target="_blank"
+          rel="noreferrer"
```
This is primarily to jailbreak out of via
